From 66beeec8488dfe0cfdf6da201ebe81d4d9aad58a Mon Sep 17 00:00:00 2001
From: Demi Marie Obenour <demi@invisiblethingslab.com>
Date: Tue, 23 Mar 2021 11:45:43 -0400
Subject: [PATCH 05/22] Avoid integer overflow computing blob->rdl

This is CVE-2021-20249.
---
 lib/header.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/header.c b/lib/header.c
index fc44070..a1b69d2 100644
--- a/lib/header.c
+++ b/lib/header.c
@@ -1837,7 +1837,8 @@ static rpmRC hdrblobVerifyRegion(rpmTagVal regionTag, int exact_size,
     }
 
     /* Is the trailer within the data area? */
-    if (hdrchkRange(blob->dl, einfo.offset + REGION_TAG_COUNT)) {
+    if (hdrchkData(einfo.offset) ||
+	hdrchkRange(blob->dl, einfo.offset + REGION_TAG_COUNT)) {
 	rasprintf(buf,
 		_("region offset: BAD, tag %d type %d offset %d count %d"),
 		einfo.tag, einfo.type, einfo.offset, einfo.count);
-- 
2.30.2

